torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -509,28 +510,26 @@ void ProcessGroupAgent::pollTimedOutRPCs() {
       return;
     }
 
-    const auto timedOutFutures = processTimedOutFutures();
+    const auto timedOutFutures = processTimedOutFutures(std::move(lock));
 
-    // Do not hold the lock while marking futures completed, as markCompleted()
-    // could invoke callbacks.
-    lock.unlock();
     for (const auto& timedOutFuture : timedOutFutures) {
       std::ostringstream ss;
       ss << "RPC ran for more than " << timedOutFuture.timeout_.count()
          << " milliseconds and timed out.";
       const auto exceptionMsg = createExceptionResponse(
           Message({}, {}, MessageType::EXCEPTION), ss.str());
+      // Do not hold the lock while marking futures completed, as
+      // markCompleted() could invoke callbacks.
       timedOutFuture.future_->markCompleted(exceptionMsg);
 
       const int dst = timedOutFuture.dstRank_;
       recvCounts_.increment(dst);
-      futureCV_.notify_all();
```

Comment:
I kept this notify_all to be in line with what's done when we process received messages. Unsure if removing this here and doing a notify after processing all timed out futures will cause correctness issues.

Reply:
Should move that up to before `fm->markCompleted(std::move(message));` too.